- Inject this.app when calling onSchema method

LobeTia opened the pull request:
Inject app when calling onSchema, needed for jaumard/trailpack-passport#77

LobeTia committed:
- Inject this.app when calling onSchema method

jaumard commented on index.js:
```diff
@@ -62,7 +62,7 @@ module.exports = class MongooseTrailpack extends DatastoreTrailpack {
         schema.methods = model.methods
 
-        model.onSchema(schema)
+        model.onSchema(schema, this.app)
```
I think it the other way around, first the app then the instance

jaumard replied:
See https://github.com/trailsjs/trailpack-sequelize/blob/master/lib/transformer.js#L14

LobeTia replied:
It was my first idea but I was worried to break the retro-compatibility for other users.
If for you there's no problem I can proceed

jaumard replied:
I prefer breaking retro compatibility and have unify interfaces between ORM. We'll update the module with a minor version not a patch. If we don't we will have if(mongoose) invert stuff in all trailpack that want to be multi ORM.

konstantinzolotarev replied:
@jaumard @LobeTia  Basically we just released version 2. So shouldn't have compatibility issues.
Might be it would be better to create unified interfaces for such things ?
And make all `datastore` trailpacks to use them ? It will solve such issues in future

jaumard replied:
Right ! Good idea @konstantinzolotarev trailpack-datastore can have this interface to prevent futur problem, because maybe other ORM trailpack already have this problem too

LobeTia replied:
Good for me

A quick note: All datastore has 2 methods: `config(app, Store)` & `schema(app Store)` that return objects.
trailpack-mongoose has this third method called onSchema for adding additionals configuration that must be a function called during bootstrap fase (essentially for adding hooks)

Why not move this into `config(app, Store)` for keeping things unified?

konstantinzolotarev replied:
@LobeTia It's made like this because of mongoose.
We couldn't add `pre/post` hooks or `virtual` fields before out schema is created. 
So we have to create schema first and after this give users an ability to manipulate with it before `Model` created from it.

LobeTia replied:
I see.
And why not allowing users to set a key named `onSchema` from `config(app, Store)` with a function as a value? We can call it after Model creation.
Like `association(model)` in trailpack-sequelize that it's called for handling associations after Model creation but returned by `config(app, Store)`

konstantinzolotarev replied:
@LobeTia it would be very dirty solution.
Because `onSchema` key in `schema()` method will confuse a lot. It's much easier to add new method and all that is related to `mongoose` would be in this method. I think it much better rather than new special key in `schema()`

LobeTia replied:
Sorry, I think I didn't explained well
I'm talking about moving the onSchema method in `config` object, somethings like this:

```
'use strict'
...
module.exports = {
  config:   (app, Mongoose) => {
    return {
      schema:   {
        timestamps: true,
        versionKey: false
      },
      methods:  {
        toJSON: function() {
          const user = this.toObject()
          user.id    = user._id
          delete user.password

          return user
        }
      },
      // REFACTORED HERE
      onSchema: function(app, schema) {
        schema.pre('save', function(next) {
          hashPassword(app.config.passport.bcrypt, this, next)
        })
      }
    }
  },
  schema: (app, Mongoose){
    ...
  }
}
```

LobeTia replied:
Hello! Any news on this?
Can I proceed with the edit moving onSchema into config method?

I'm quite stuck for jaumard/trailpack-passport#77 without progress on this
@konstantinzolotarev @jaumard

jaumard reviewed (changes requested):
Inverse attributs to be compliant with other trailpack ORM

konstantinzolotarev commented:
Yep if you want. I'm a bit busy for now.
So if you are able to do it it would be awesome

On Feb 2, 2017 19:08, "Mattia Lobertini" <notifications@github.com> wrote:

> *@LobeTia* commented on this pull request.
> ------------------------------
>
> In index.js <https://github.com/trailsjs/trailpack-mongoose/pull/28>:
>
> > @@ -62,7 +62,7 @@ module.exports = class MongooseTrailpack extends DatastoreTrailpack {
>          schema.statics = model.statics
>          schema.methods = model.methods
>
> -        model.onSchema(schema)
> +        model.onSchema(schema, this.app)
>
> Hello! Any news on this?
> Can I proceed with the edit moving onSchema into config method?
>
> I'm quite stuck for jaumard/trailpack-passport#77
> <https://github.com/jaumard/trailpack-passport/issues/77> without
> progress on this
> @konstantinzolotarev <https://github.com/konstantinzolotarev> @jaumard
> <https://github.com/jaumard>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/trailsjs/trailpack-mongoose/pull/28>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AARe32KE8gsgU-3OsNjmblnfRFgYR-LEks5rYg2PgaJpZM4LqH9a>
> .
>

LobeTia committed:
- Fixed README

LobeTia committed:
- onSchema refactored inside config(app, Mongoose)

LobeTia commented:
Hello!
I refactored `onSchema` inside of `config(app, Mongoose)` ad inverted attributes of `onSchema` signature (first app, then schema)
I also added `index.test.js` for checking coverage and updated `README.md` for reflecting this changes

I think it's ok, maybe we can review it and release? @jaumard @konstantinzolotarev 
**Warning: this is a breaking change from the past `onSchema` behavior, semver should be updated properly for avoiding problems**

LobeTia committed:
- Fixed some README typo

konstantinzolotarev commented:
@LobeTia Thank you very much !
